src/libmeasurement_kit/dns/system_resolver.hpp

```
@@ -0,0 +1,141 @@
+                    Var<Logger> l) {
+        dns_class = dns_c;
+        dns_type = dns_t;
+        name = n;
+        cb = c;
+        settings = s;
+        reactor = r;
+        logger = l;
+        memset(&hints, 0, sizeof(hints));
+    }
+
+    ~ResolverContext() {
+        if (servinfo != nullptr) {
+            freeaddrinfo(servinfo);
+        }
+    }
+};
+
+template<MK_MOCK(getaddrinfo), MK_MOCK(inet_ntop)>
+static inline void resolve_async(ResolverContext *context) {
+    std::unique_ptr<ResolverContext> ctx(context);
+    Callback<Error, Var<Message>> callback = ctx->cb;
+    Var<Message> message_copy;
+
+    int error = getaddrinfo(ctx->name.c_str(), nullptr, &(ctx->hints),
+                            &(ctx->servinfo));
+    if (error) {
+        // check the error variable and return the correct error
+        ctx->logger->warn(gai_strerror(error));
+        ctx->reactor->call_soon(
+            [callback]() { callback(ResolverError(), nullptr); });
+        return;
+    }
+
+    assert(ctx->servinfo != nullptr);
+
+    void *addr_ptr;
+    char address[128];
+    std::vector<Answer> answers;
+    for (addrinfo *p = ctx->servinfo; p != nullptr; p = p->ai_next) {
+        Answer answer;
+        answer.name = ctx->name;
+        answer.qclass = ctx->dns_class;
+        if (p->ai_family == AF_INET) {
+            answer.type = QueryTypeId::A;
+            addr_ptr = &((sockaddr_in *)p->ai_addr)->sin_addr;
+        } else if (p->ai_family == AF_INET6) {
+            answer.type = QueryTypeId::AAAA;
+            addr_ptr = &((sockaddr_in6 *)p->ai_addr)->sin6_addr;
+        } else {
+            ctx->reactor->call_soon(
+                [callback]() { callback(ResolverError(), nullptr); });
+            return;
+        }
+        if (inet_ntop(p->ai_family, addr_ptr, address, sizeof(address)) ==
+            NULL) {
+            ctx->logger->warn("dns: unexpected inet_ntop failure");
+            throw std::runtime_error("Unexpected inet_ntop failure");
```

> We do this also in the libevent dns resolver function, so we should change it also there.

Yes. I remember that the argument for throwing was "it should not happen". However, if something goes wrong and a mobile app crashes because we've thrown it's a pain to debug it. So, probably it would be better to return an error. On the other hands, things that really should not happen (e.g. you read up to 1,024 bytes and the result is 2,048 read bytes) should be assertions.